# SaCas9_off-target_activity
with https://github.com/editasmedicine/pairwise-library-screen

1. find off-target candidate by Cas-Offinder
    result form : 
        WT(guide)  chromosome  location    mis_match_seq   strand  #_of_mis_match

2-1. filter out if #_mis_match >= 2  ==> list of filterd seq by guide(21 bp)
2-2. filter out if #_mis_match == 0 doesn't exist ==> list of seq that #_mis_match == 0 doesn't exist
    result :
        WT(guide)[:21]   #_of_mis_match  #_seq  tot_seq
        AGCT...CGT  0   2   15
        AGCT...AGT  0   1   8

2. get off-target score by predict_activity_single
    input : WT[:21] mis_match_seq[:21]
    
    

####################################################################################    
1) guide RNAs design
    1-1) Initial input file: 201130_CCDS_human_current.txt, 201130_CCDS_mouse_current.txt
    1-2) Input filter
        1-2-1) ccds_id: ê¸¸ì´ê°€ ê°€ìž¥ ì§§ì€ cdsì— í•´ë‹¹í•˜ëŠ” ccds_idë§Œ ì‚¬ìš©
        1-2-2) ccds_status: Publicë§Œ ì‚¬ìš©
        1-2-3) match_type: Identicalë§Œ ì‚¬ìš©
    1-3) Reference genome (2200): /media/backup/ref/Ensemble_GRCh38_p13, /media/backup/ref/Ensemble_GRCm38_p6
    1-4) guide RNA ê¸°ì¤€
        1-4-1) cdsì˜ 5-65% êµ¬ê°„ì— cleavage siteê°€ ìœ„ì¹˜í•˜ëŠ” guide RNA contextë¥¼ genomic regionì—ì„œ ì°¾ìŒ (cleavage siteê°€ junctionì— ê²¹ì¹˜ë©´ ì œê±°)
        1-4-2) Cas9ë³„ë¡œ ê¸¸ì´ ë° PAM ì¡°ê±´ì€ ê¸°ì¡´ì— ì „ë‹¬í•´ë“œë¦° ê²ƒê³¼ ê°™ì•„ ë”°ë¡œ ì ì§€ ì•Šì•˜ìœ¼ë‚˜ í•„ìš”í•˜ì‹œë©´ ì´ íŒŒì¼ì— ì—…ë°ì´íŠ¸ í•˜ê² ìŠµë‹ˆë‹¤.
    1-5)  Output file
        1-5-1) Filtered input
        1-5-2) sgRNA with context sequence (location, strand, cds ë‚´ì—ì„œ cleavage siteì˜ ìœ„ì¹˜ ì •ë³´ ë“±ì„ í¬í•¨)

2) Cas-offinder (ì œê°€ ëŒë¦´ ë¶€ë¶„ì´ë¼ ìƒëžµí•˜ê² ìŠµë‹ˆë‹¤.)
    
3) SaCas9 off-target activity
    3-1) Initial input file directory: B206 /extdata1/JaeWoo/Project/YoungGwang_CasOffFinder/Output/FirstResult/ 
    3-2) Soff-target: [a relative link](./pairwise-library-screen-master/predict_activity_single.py) https://github.com/editasmedicine/pairwise-library-screen ì˜ predict_activity_single.py
    3-3) Off-target rank
        3-3-1) Tier: off-target positionì´ CDS (Tier I)ì¸ì§€ non-CDS (Tier II)ì¸ì§€ ë¶„ë¥˜
            3-3-1-1) Inputì€ Cas-OFFinder ê²°ê³¼ì˜ 2ë²ˆì§¸ columnì—ì„œ chromosome ì •ë³´ (Cas-OFFinder output fileì´ ì¡°ê¸ˆ ì§€ì €ë¶„í•´ chromosome ì •ë³´ë¥¼ ê°€ì ¸ì˜¬ ë•Œ í™•ì¸ì´ í•„ìš”í•  ê²ƒ ê°™ìŠµë‹ˆë‹¤)
            3-3-1-2) 5ë²ˆì§¸ columnì´ +ë©´, 3ë²ˆì§¸ columnì˜ position + 19ê°€ cleavage site (ì„ ìƒë‹˜ì´ get_seq_near_target_indexë¥¼ í•´ì£¼ì‹¤ ë•Œì˜ position ë°©ì‹ê³¼ ê°™ê²Œ í•˜ë ¤ê³  ìœ„ì¹˜ ì •ë³´ë¥¼ ì¼ë¶€ ìˆ˜ì •í•˜ëŠ” ê³¼ì •ìž…ë‹ˆë‹¤)
            3-3-1-3) 5ë²ˆì§¸ columnì´ â€“ë©´, 3ë²ˆì§¸ columnì˜ position + 10ê°€ cleavage site (ì„ ìƒë‹˜ì´ get_seq_near_target_indexë¥¼ í•´ì£¼ì‹¤ ë•Œì˜ position ë°©ì‹ê³¼ ê°™ê²Œ í•˜ë ¤ê³  ìœ„ì¹˜ ì •ë³´ë¥¼ ì¼ë¶€ ìˆ˜ì •í•˜ëŠ” ê³¼ì •ìž…ë‹ˆë‹¤)
            3-3-1-4) CDS referenceëŠ” 1-2-1ì¡°ê±´ì„ ì œì™¸í•œ ccds fileì„ ì‚¬ìš©, ì¦‰ í•˜ë‚˜ì˜ geneì— ëŒ€í•´ì„œ ëª¨ë“  transcript) ==> make_filtered_ccds_current_file_by_all_ccds_id() [a relative link](./MakeCDSInput.py)
        3-3-2) Match bin: Soff-targetì„ 0.05, 0.2, 1ì„ ê¸°ì¤€ìœ¼ë¡œ ë¶„ë¥˜, ì¦‰ Soff-target = 1 (Match bin I), 1 > Soff-target  >= 0.2 (Match bin II), 0.2 > Soff-target >= 0.05 (Match bin III), Soff-target < 0.05 (Match bin IV)ìœ¼ë¡œ ë¶„ë¥˜
    3-4) Sguide = 100/(1+ âˆ‘_(ð‘–=1)^ð‘›â–’ã€–ð‘†ð‘œð‘“ð‘“âˆ’ð‘¡ð‘Žð‘Ÿð‘”ð‘’ð‘¡(ð‘–)ã€—)
    
    3-5) Output file
    ![alt text](./result.PNG)
